File: android/app/src/main/java/org/bitcoinppl/cove/flows/keyteleport/KeyTeleportReceiveViews.kt
package org.bitcoinppl.cove.flows.keyteleport

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.QrCodeScanner
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.bitcoinppl.cove.flows.OnboardingFlow.OnboardingPrimaryButton
import org.bitcoinppl.cove.flows.OnboardingFlow.OnboardingStatusHero
import org.bitcoinppl.cove.flows.OnboardingFlow.OnboardingTextSecondary
import org.bitcoinppl.cove_core.KeyTeleportManagerAction
import org.bitcoinppl.cove_core.KeyTeleportReceiveState
import org.bitcoinppl.cove_core.WalletMetadata

private val ImportedWalletSuccessTint = Color(0xFF7DD195)
private val ImportedWalletSuccessFill = Color(0x297DD195)

internal enum class ImportedWalletStatus {
    IMPORTED,
    ALREADY_IMPORTED,
}

private data class ImportedWalletContent(
    val title: String,
    val message: String,
    val buttonTitle: String,
)

@Composable
internal fun ReceiveReadyView(
    receive: KeyTeleportReceiveState,
    concealmentGeneration: Long,
    onScan: () -> Unit,
) {
    val packetText = remember(receive.packet) { runCatching { receive.packet.bbqrPart() }.getOrNull() }

    SecureScreenEffect()
    if (packetText == null) {
        Text("Unable to render this receive request.", color = MaterialTheme.colorScheme.error)
        ReceiverCode(receive.groupedNumericCode)
    } else {
        KeyTeleportRevealPair(
            qrHint = "Tap to show QR code",
            codeHint = "Tap to show receiver code",
            resetKey = concealmentGeneration,
            qr = { PacketQr(packetText) },
            code = { ReceiverCode(receive.groupedNumericCode) },
        )
    }
    Text(
        text =
            "Have the sending wallet scan the QR code, then send the receiver code through a different " +
                "channel, such as a call or message.\n\n" +
                "If the sending wallet cannot scan this screen, tap Share and open the link on another " +
                "device. The link shows the same QR code.",
        color = OnboardingTextSecondary,
        style = MaterialTheme.typography.bodySmall,
    )
    OnboardingPrimaryButton(
        text = "Scan Sender Response",
        onClick = onScan,
        icon = Icons.Default.QrCodeScanner,
    )
}

@Composable
private fun ReceiverCode(code: String) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Text(
            text = "Receiver Code",
            color = OnboardingTextSecondary,
            style = MaterialTheme.typography.labelMedium,
        )
        KeyTeleportCodeText(code)
    }
}

@Composable
internal fun ReceivePasswordView(manager: KeyTeleportManager) {
    var password by remember { mutableStateOf("") }

    SecureScreenEffect()
    TextBlock(
        title = "Enter sender password",
        body = "Type the password shown by the sending device.",
    )
    OutlinedTextField(
        value = password,
        onValueChange = { password = it },
        label = { Text("Password") },
        singleLine = true,
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
        visualTransformation = PasswordVisualTransformation(),
        colors = keyTeleportTextFieldColors(),
        modifier = Modifier.fillMaxWidth(),
    )
    Button(
        enabled = password.isNotBlank(),
        onClick = { manager.dispatch(KeyTeleportManagerAction.EnterSenderPassword(password.trim())) },
        modifier = Modifier.fillMaxWidth(),
    ) {
        Text("Continue")
    }
}

@Composable
internal fun ReceiveImportedWalletView(
    manager: KeyTeleportManager,
    wallet: WalletMetadata,
    status: ImportedWalletStatus,
    onDone: () -> Unit,
) {
    val content =
        when (status) {
            ImportedWalletStatus.IMPORTED ->
                ImportedWalletContent(
                    title = "Wallet imported",
                    message = "${wallet.name} is ready to use in Cove.",
                    buttonTitle = "Done",
                )

            ImportedWalletStatus.ALREADY_IMPORTED ->
                ImportedWalletContent(
                    title = "Wallet already imported",
                    message = "${wallet.name} is already available in Cove.",
                    buttonTitle = "Open Wallet",
                )
        }

    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(18.dp),
    ) {
        OnboardingStatusHero(
            icon = Icons.Default.Check,
            tint = ImportedWalletSuccessTint,
            fillColor = ImportedWalletSuccessFill,
        )
        Text(content.title, color = Color.White, fontSize = 26.sp, fontWeight = FontWeight.SemiBold)
        Text(content.message, color = OnboardingTextSecondary)
        Button(
            onClick = {
                manager.dispatch(KeyTeleportManagerAction.Clear)
                onDone()
            },
            modifier = Modifier.fillMaxWidth(),
        ) {
            Text(content.buttonTitle)
        }
    }
}

File: ios/Cove/Flows/KeyTeleportFlow/KeyTeleportReceiveViews.swift
import SwiftUI

struct KeyTeleportReceiveSetupStateContent: View {
    @Bindable var manager: KeyTeleportManager
    @Binding var senderPassword: String
    let scan: () -> Void

    var body: some View {
        switch manager.state {
        case .idle:
            KeyTeleportLoadingView()
        case let .receiveReady(state):
            KeyTeleportReceiveReadyView(state: state, scan: scan)
        case .receiveError:
            KeyTeleportReceiveErrorView(retry: retry)
        case .receiveEnterPassword:
            KeyTeleportPasswordEntryView(password: $senderPassword, submit: submitPassword)
        default:
            EmptyView()
        }
    }

    private func retry() {
        manager.dispatch(.startReceive)
    }

    private func submitPassword() {
        manager.dispatch(.enterSenderPassword(senderPassword))
    }
}

struct KeyTeleportReceiveErrorView: View {
    let retry: () -> Void

    var body: some View {
        VStack(spacing: 16) {
            Text("Cove couldn’t prepare a receive request.")
                .font(.headline)

            Button("Try Again", action: retry)
                .buttonStyle(OnboardingPrimaryButtonStyle())
        }
    }
}

struct KeyTeleportReceiveReviewStateContent: View {
    @Bindable var manager: KeyTeleportManager
    @Binding var mnemonicDisclosure: KeyTeleportMnemonicDisclosure
    @Binding var xprv: String?
    let revealMnemonicWords: () -> Void
    let finishReview: () -> Void

    var body: some View {
        switch manager.state {
        case let .receiveMnemonicReview(review):
            KeyTeleportMnemonicReviewView(
                review: review,
                disclosure: mnemonicDisclosure,
                reveal: revealMnemonicWords,
                importWords: importMnemonic,
                finish: finishMnemonicReview
            )
        case let .receiveXprvReview(review):
            KeyTeleportXprvReviewView(
                review: review,
                xprv: $xprv,
                reveal: revealXprv,
                hide: hideXprv,
                importWallet: importXprv,
                finish: finishXprvReview
            )
        case let .receiveMessageReview(review):
            KeyTeleportMessageReviewView(review: review, finish: finishReview)
        case let .receiveImportedWallet(wallet):
            KeyTeleportImportedWalletStateView(
                manager: manager,
                wallet: wallet,
                presentation: .imported
            )
        case let .receiveAlreadyImportedWallet(wallet):
            KeyTeleportImportedWalletStateView(
                manager: manager,
                wallet: wallet,
                presentation: .alreadyImported
            )
        default:
            EmptyView()
        }
    }

    private func importMnemonic() {
        mnemonicDisclosure = .hidden
        manager.dispatch(.importReceivedWallet)
    }

    private func finishMnemonicReview() {
        mnemonicDisclosure = .hidden
        finishReview()
    }

    private func revealXprv() {
        xprv = manager.revealXprv()
    }

    private func hideXprv() {
        xprv = nil
        manager.dispatch(.hideXprv)
    }

    private func importXprv() {
        xprv = nil
        manager.dispatch(.importReceivedWallet)
    }

    private func finishXprvReview() {
        xprv = nil
        finishReview()
    }
}

enum KeyTeleportImportedWalletPresentation {
    case imported
    case alreadyImported
}

struct KeyTeleportImportedWalletStateView: View {
    @Environment(AppManager.self) private var app

    @Bindable var manager: KeyTeleportManager
    let wallet: WalletMetadata
    let presentation: KeyTeleportImportedWalletPresentation

    var body: some View {
        switch presentation {
        case .imported:
            KeyTeleportImportedWalletView(wallet: wallet, finish: openWallet)
        case .alreadyImported:
            KeyTeleportImportedWalletView(
                wallet: wallet,
                title: "Wallet already imported",
                message: "\(wallet.name) is already available in Cove.",
                buttonTitle: "Open Wallet",
                finish: openWallet
            )
        }
    }

    private func openWallet() {
        manager.dispatch(.clear)
        app.selectWallet(wallet.id)
    }
}

struct KeyTeleportReceiveReadyView: View {
    let state: KeyTeleportReceiveState
    let scan: () -> Void

    var body: some View {
        VStack(spacing: 18) {
            if let packet = try? state.packet.bbqrPart() {
                KeyTeleportRevealPair(
                    qrHint: "Tap to show QR code",
                    codeHint: "Tap to show receiver code"
                ) {
                    QrCodeView(text: packet)
                        .frame(maxWidth: 280)
                        .frame(maxWidth: .infinity)
                } code: {
                    receiverCode
                }
            } else {
                Text("Unable to render this receive request.")
                    .foregroundStyle(.red)

                receiverCode
            }

            Text(
                """
                Have the sending wallet scan the QR code, then send the receiver code through a different channel, such as a call or message.

                If the sending wallet cannot scan this screen, tap Share and open the link on another device. The link shows the same QR code.
                """
            )
            .font(.caption)
            .foregroundStyle(.secondary)
            .fixedSize(horizontal: false, vertical: true)

            Button(action: scan) {
                Label("Scan Sender Response", systemImage: "qrcode.viewfinder")
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(OnboardingPrimaryButtonStyle())
        }
    }

    private var receiverCode: some View {
        VStack(spacing: 4) {
            Text("Receiver Code")
                .font(.caption)
                .foregroundStyle(.secondary)
            KeyTeleportCodeText(state.groupedNumericCode)
        }
        .frame(maxWidth: .infinity)
    }
}

struct KeyTeleportPasswordEntryView: View {
    @Binding var password: String
    let submit: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            KeyTeleportSecureInput(text: $password, submit: submit)

            Button(action: submit) {
                Text("Unlock")
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(OnboardingPrimaryButtonStyle())
        }
    }
}

struct KeyTeleportImportedWalletView: View {
    let wallet: WalletMetadata
    var title = "Wallet imported"
    var message: String?
    var buttonTitle = "Done"
    let finish: () -> Void

    var body: some View {
        VStack(spacing: 18) {
            OnboardingStatusHero(
                systemImage: "checkmark",
                tint: .green,
                fillColor: .green.opacity(0.16),
                iconSize: 22,
                innerBadgeSize: 58
            )

            VStack(spacing: 8) {
                Text(title)
                    .font(OnboardingRecoveryTypography.compactTitle)

                Text(message ?? "\(wallet.name) is ready to use in Cove.")
                    .font(.subheadline)
                    .foregroundStyle(.coveLightGray.opacity(0.74))
                    .multilineTextAlignment(.center)
            }

            Button(buttonTitle, action: finish)
                .buttonStyle(OnboardingPrimaryButtonStyle())
        }
        .frame(maxWidth: .infinity)
    }
}
